crypto/argon2/argon2.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Argon2 reference source code package - reference C implementations
+ *
+ * Copyright 2015
+ * Daniel Dinu, Dmitry Khovratovich, Jean-Philippe Aumasson, and Samuel Neves
+ *
+ * You may use this work under the terms of a Creative Commons CC0 1.0
+ * License/Waiver or the Apache Public License 2.0, at your option. The terms of
+ * these licenses can be found at:
+ *
+ * - CC0 1.0 Universal : http://creativecommons.org/publicdomain/zero/1.0
+ * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * You should have received a copy of both of these licenses along with this
+ * software. If not, they may be obtained at the above URLs.
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include <openssl/crypto.h>
+#include <internal/blake2.h>
+
+#include "argon2.h"
+#include "core.h"
+
+typedef struct argon2_param ARGON2_PARAM;
+
+int ossl_alloc(uint8_t **memory, size_t bytes)
+{
+    if (memory == NULL)
+	return -1;
+
+    *memory = OPENSSL_zalloc(bytes);
+
+    return memory != NULL;
+}
+
+void ossl_dealloc(uint8_t *memory, size_t bytes)
+{
+    if (bytes)
+	OPENSSL_clear_free(memory, bytes);
+    else
+	OPENSSL_free(memory);
+}
+
+int ARGON2_Init(ARGON2_CTX *c, argon2_type type)
+{
+    c->out = NULL;
```

Comment:
a `memset(c, 0, sizeof(*c))` might be a better way to do this, followed by setting the non-zero fields.